hngprojects/hng_boilerplate_java_web
# feat: create roles and permissions within organisation

## How should this be manually tested?
- Try accessing the `/api/v1/organisations` route as an authorized user (Bearer Token).
- Submit a valid PATCH request to update an organisation.
- Submit a valid GET request to `/me` path to get list of organisations a user belongs in
- Submit a valid GET request to `/:orgId/users` to fetch all users in an organisation
- Submit a valid GET request to `/:orgId` to find an organisation by ID
- The expected result is to get a `200 OK` response with the organisation details.

- Try accessing the `/api/v1/organisation` route as an authorized user (Bearer Token).
- Submit a valid POST request to `/:orgId/permissions' to create a new permission
- Submit a valid POST request to `/:orgId/roles` to create a new role
- The expected result is to get a `201 CREATED` response with the created role or permission details.

- Error responses to check:
  - `403 Unauthorized` if the user is not authenticated.
  - `422 Unprocessable Entity` if there are validation errors.
  -  `400 Bad Request` if the organisation name already exists.
  -  `400 Bad Request` if the role name already exists.
  -  `400 Bad Request` if the permission name already exists.

## Checklist of what you did
- [x] Implemented authentication middleware to ensure the user is authenticated using JWT.
- [x] Create the `/api/v1/organisations` endpoint to handle organisation update.
- [x] Create the `/api/v1/organisations/me` endpoint to fetch organisations the user belongs in.
- [x] Create the `/api/v1/organisations/:orgId/users` to fetch all users in an organisation.
- [x] Create the `/api/v1/organisations/:orgId` to find an organisation by ID.
- [x] Create the `/api/v1/organisation/:orgId/permissions` endpoint to create permission.
- [x] Create the `/api/v1/organisations/:orgId/roles` endpoint to create role

- [x] Added input validation for all supplied fields in the request body.
- [x] Implemented error handling for various scenarios including name conflicts, validation errors, unauthorized access, and internal server errors.
- [x] Ensured the response structure is consistent for successful and error scenarios.
- [x] Added unit tests for all the features as described more in the reference issue below.

## Reference Issues
[https://github.com/hngprojects/hng_boilerplate_nestjs/issues/259]()
[https://github.com/hngprojects/hng_boilerplate_nestjs/issues/166]()
